InventoryTweaks/Patches/Helpers.cs

```diff
@@ -0,0 +1,63 @@
+﻿using HarmonyLib;
+
+// ReSharper disable InconsistentNaming
+
+namespace thmsn.InventoryTweaks.Patches
+{
+    internal static class Helpers
+    {
+        private static readonly AccessTools.FieldRef<Slot, Inventory> Slot_inventory =
+            AccessTools.FieldRefAccess<Slot, Inventory>(AccessTools.Field(typeof(Slot), "inventory"));
+
+        private static readonly FastInvokeHandler PlayerInventory_MoveSlotToEmpty =
+            MethodInvoker.GetHandler(AccessTools.Method(typeof(PlayerInventory), "MoveSlotToEmpty"));
+
+        private static readonly FastInvokeHandler PlayerInventory_SwitchSlots =
+            MethodInvoker.GetHandler(AccessTools.Method(typeof(PlayerInventory), "SwitchSlots"));
+
+        // Faster, but needs Harmony 2.0.2
+
+        // public delegate void MoveSlotToEmpty(Inventory inventory, Slot fromSlot, Slot toSlot, int amount);
+        // public delegate void SwitchSlots(Inventory inventory, Slot fromSlot, Slot toSlot);
+        //
+        // private static readonly MoveSlotToEmpty MoveSlotToEmptyDelegate =
+        //     AccessTools.MethodDelegate<MoveSlotToEmpty>(AccessTools.Method(typeof(PlayerInventory), "MoveSlotToEmpty"));
+        // private static readonly SwitchSlots SwitchSlotsDelegate =
+        //     AccessTools.MethodDelegate<SwitchSlots>(AccessTools.Method(typeof(PlayerInventory), "SwitchSlots"));
+
+        internal static bool IsSelectedHotbarSlot(Slot slot)
+        {
+            var inventory = Slot_inventory(slot) as PlayerInventory;
+            return inventory != null && slot == inventory.GetSelectedHotbarSlot();
+        }
+
+        internal static void RefillSlotIfNeeded(Slot slot, int originalUniqueIndex,
+            PlayerInventory playerInventory = null)
+        {
+            if (slot.itemInstance?.UniqueIndex == originalUniqueIndex)
+                return;
+
+            var inventory = playerInventory ? playerInventory : Slot_inventory(slot) as PlayerInventory;
+            if (inventory is null)
+                return;
+
+            foreach (var localSlot in inventory.allSlots)
+            {
+                // last condition will never be triggered with second, but let it be
+                if (localSlot.IsEmpty || localSlot.slotType == SlotType.Hotbar || localSlot == slot)
+                    continue;
+
+                var slotItemInstance = localSlot.itemInstance;
+                if (slotItemInstance.UniqueIndex != originalUniqueIndex)
+                    continue;
+
+                if (slot.IsEmpty)
+                    PlayerInventory_MoveSlotToEmpty(inventory,
+                        new object[] { localSlot, slot, slotItemInstance.Amount });
+                else
+                    PlayerInventory_SwitchSlots(inventory, new object[] { localSlot, slot });

```

Comment:
Does this still remove the old item, if we swap the "broken" item with a new one from the inventory?

Reply:
It's only for items with empty state, like water bottle